src/eos-util.c

```diff
@@ -421,3 +452,142 @@ eos_updater_get_timestamp_from_branch_file_keyfile (GKeyFile *branch_file,
   *out_timestamp = g_steal_pointer (&timestamp);
   return TRUE;
 }
+
+gchar *
+eos_updater_dup_envvar_or (const gchar *envvar,
+                           const gchar *default_value)
+{
+  const gchar *value = g_getenv (envvar);
+
+  if (value != NULL)
+    return g_strdup (value);
+
+  return g_strdup (default_value);
+}
+
+struct _EosQuitFile
+{
+  GObject parent_instance;
+
+  GFileMonitor *monitor;
+  guint signal_id;
+  guint timeout_seconds;
+  guint timeout_id;
+  EosQuitFileCheckCallback callback;
+  gpointer user_data;
+  GDestroyNotify notify;
+};
+
+static void
+quit_clear_user_data (EosQuitFile *quit_file)
+{
+  gpointer user_data = g_steal_pointer (&quit_file->user_data);
+  GDestroyNotify notify = g_steal_pointer (&quit_file->notify);
+
+  if (notify != NULL)
+    notify (user_data);
+}
+
+static void
+quit_disconnect_monitor (EosQuitFile *quit_file)
+{
+  guint id = quit_file->signal_id;
+
+  quit_file->signal_id = 0;
+  if (id > 0)
+    g_signal_handler_disconnect (quit_file->monitor, id);
+}
+
+static void
+quit_clear_source (EosQuitFile *quit_file)
+{
+  guint id = quit_file->timeout_id;
+
+  quit_file->timeout_id = 0;
+  if (id > 0)
+    g_source_remove (id);
+}
+
+static void
+eos_quit_file_dispose_impl (EosQuitFile *quit_file)
+{
+  quit_clear_user_data (quit_file);
+  quit_clear_source (quit_file);
+  quit_disconnect_monitor (quit_file);
+  g_clear_object (&quit_file->monitor);
+}
+
+EOS_DEFINE_REFCOUNTED (EOS_QUIT_FILE,
+                       EosQuitFile,
+                       eos_quit_file,
+                       eos_quit_file_dispose_impl,
+                       NULL)
+
+static gboolean
+quit_file_source_func (gpointer quit_file_ptr)
+{
+  EosQuitFile *quit_file = EOS_QUIT_FILE (quit_file_ptr);
+
+  if (quit_file->callback (quit_file->user_data) == EOS_QUIT_FILE_KEEP_CHECKING)
+    return G_SOURCE_CONTINUE;
+
+  quit_file->timeout_id = 0;
+  quit_clear_user_data (quit_file);
+  return G_SOURCE_REMOVE;
+}
+
+static void
+on_quit_file_changed (GFileMonitor *monitor,
+                      GFile *file,
+                      GFile *other,
+                      GFileMonitorEvent event,
+                      gpointer quit_file_ptr)
+{
+  EosQuitFile *quit_file = EOS_QUIT_FILE (quit_file_ptr);
+  guint id;
+
+  if (event != G_FILE_MONITOR_EVENT_DELETED)
+    return;
+
+  if (quit_file->callback (quit_file->user_data) == EOS_QUIT_FILE_KEEP_CHECKING)
+    quit_file->timeout_id = g_timeout_add_seconds (quit_file->timeout_seconds,
+                                                   quit_file_source_func,
+                                                   quit_file);
+  id = quit_file->signal_id;
+  quit_file->signal_id = 0;
+  g_signal_handler_disconnect (quit_file->monitor, id);
```

Comment:
Could also be written as:

    g_signal_handler_disconnect (quit_file->monitor, quit_file->signal_id);
    quit_file->signal_id;